### DISCUSSION TITLE
Custom ValidationError with client response

### DISCUSSION THREAD

**Discussion by brunolnetto** [pydantic/pydantic#10749] [Feature Request]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

I am using pydantic as usual for Request-Response models. One of my query parameters is a date. For that, I pass the value as a string, since I do not know how to do otherwise. Take a look at my attempt below. To run it, paste to a file `main.py` and run command (on Linux or OS) `uvicorn main:app --reload --host 0.0.0.0 --port 8000`

```py
from typing import Annotated
from datetime import datetime

from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel, Field, field_validator

app = FastAPI()

class MyModel(BaseModel):
    my_date: str = Field(
        None, description="The date for a model.")

    @field_validator("my_date")
    def parse_date(cls, value):
        if isinstance(value, str):
            for fmt in ("%d/%m/%Y", "%Y-%m-%d"):
                try:
                    return datetime.strptime(value, fmt).date()
                except ValueError:
                    continue
            # Raise an error if neither format worked
            raise ValueError("Date must be in 'DD/MM/YYYY' or 'YYYY-MM-DD' format")
        return value

app = FastAPI()

@app.get("/")
def read_root(my_model: MyModel = Depends()):
    return {"Hello": "World"}
```

It does raise a pydantic `ValidationError`. However, it does not appear on for the client, just on terminal, which gives an Internal error and disable the server. What do you recommend me to do?

``` 
pydantic_core._pydantic_core.ValidationError: 1 validation error for MyModel
my_date
Value error, Date must be in 'DD/MM/YYYY' or 'YYYY-MM-DD' format [type=value_error, input_value='02/02/ABC', input_type=str]
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
Hi @brunolnetto,

Seems like this is the expected behavior, as your input data isn't a valid date. Going to move this to a discussion, as I think it's more of a FastAPI setup question than a `pydantic` bug.

Perhaps @Kludex would have some insight here, my best guess would be to add a try / except block to that endpoint that catches validation errors and handles them gracefully.

**Reply by brunolnetto** [reply]
Sure. Please, may you accept my humble invitation on LinkedIN? Thanks. :)

**Comment by andryunin** [comment]
Hi @brunolnetto,

You should tell FastAPI that you model is [query parameter model](https://fastapi.tiangolo.com/tutorial/query-param-models/):

```
@app.get("/")
def read_root(my_model: Annotated[MyModel, Query()]):
    return {"Hello": "World"}
```

**Reply by brunolnetto** [reply]
Since you are here, when should I use Annotated?

**Reply by andryunin** [reply]
In general, I think, when you want to add some context to type. In case of FastAPI, it is useful to adding stuff like custom validation rules, like [Before](https://docs.pydantic.dev/latest/api/functional_validators/#pydantic.functional_validators.BeforeValidator)/[AfterValidator](https://docs.pydantic.dev/latest/api/functional_validators/#pydantic.functional_validators.AfterValidator).

Actually, just found out it works without Annotated too:

```
@app.get("/")
def read_root(my_model: MyModel = Query()):
    return {"Hello": "World"}
```

Maybe using Annotated is just more modern way to describe params, I believe in older versions of documentation Query was used like this.

**Reply by brunolnetto** [reply]
Outstanding! Thanks for the quick reply.

**Reply by brunolnetto** [reply]
@andryunin I tested this only today (my bad). It leads to error `AssertionError: Param: my_model can only be a request body, using Body()`, instead. In my head, I could define a pydantic object and use as a single Query object instead of defining one by one. is it so?
